gcc/rust/ast/rust-ast-tokenstream.cc

```diff
@@ -0,0 +1,2718 @@
+// Copyright (C) 2020-2023 Free Software Foundation, Inc.
+
+// This file is part of GCC.
+
+// GCC is free software; you can redistribute it and/or modify it under
+// the terms of the GNU General Public License as published by the Free
+// Software Foundation; either version 3, or (at your option) any later
+// version.
+
+// GCC is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+// for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with GCC; see the file COPYING3.  If not see
+// <http://www.gnu.org/licenses/>.
+#include "rust-ast-tokenstream.h"
+
+namespace Rust {
+namespace AST {
+void
+TokenStream::go (AST::Crate &crate)
+{
+  visit_items_as_lines (crate.items);
+}
+
+void
+TokenStream::go (AST::Item &item)
+{
+  item.accept_vis (*this);
+}
+
+template <typename T>
+void
+TokenStream::visit (std::unique_ptr<T> &node)
+{
+  node->accept_vis (*this);
+}
+
+template <typename T>
+void
+TokenStream::visit (T &node)
+{
+  node.accept_vis (*this);
+}
+
+template <typename T>
+void
+TokenStream::visit_items_joined_by_separator (T &collection, TokenId separator,
+					      size_t start_offset,
+					      size_t end_offset)
+{
+  if (collection.size () > start_offset)
+    {
+      visit (collection.at (start_offset));
+      auto size = collection.size () - end_offset;
+      for (size_t i = start_offset + 1; i < size; i++)
+	{
+	  tokens.push_back (Rust::Token::make (separator, Location ()));
+	  visit (collection.at (i));
+	}
+    }
+}
+
+template <typename T>
+void
+TokenStream::visit_as_line (T &item, std::vector<TokenPtr> trailing)
+{
+  visit (item);
+  for (auto &token : trailing)
+    {
+      tokens.push_back (token);
+    }
```

Comment:
I believe the GNU style prefers
```suggestion
      tokens.push_back (token);
```

but this is a *very tiny nitpick*